autopush/router/apns2.py

```diff
@@ -0,0 +1,152 @@
+import json
+import uuid
+from collections import deque
+
+import hyper.tls
+from hyper import HTTP20Connection
+from twisted.internet.threads import deferToThread
+
+SANDBOX = 'api.development.push.apple.com'
+SERVER = 'api.push.apple.com'
+
+APNS_PRIORITY_IMMEDIATE = 10
+APNS_PRIORITY_LOW = 5
+APNS_MAX_CONNECTIONS = 20
+APNS_MIN_CONNECTIONS = 5
+
+
+class APNSException(Exception):
+    pass
+
+
+class APNSClient(object):
+    def __init__(self, cert_file, key_file, topic,
+                 alt=False, use_sandbox=False,
+                 max_connections=APNS_MAX_CONNECTIONS,
+                 min_connections=APNS_MIN_CONNECTIONS,
+                 logger=None, metrics=None):
+        """Create the APNS client connector.
+
+        The cert_file and key_file can be derived from the exported `.p12`
+        **Apple Push Services: *bundleID*** key contained in the **Keychain
+        Access** application. To extract the proper PEM formatted data, you
+        can use the following commands:
+
+        ```
+        openssl pkcs12 -in file.p12 -out apns_cert.pem -clcerts -nokeys
+        openssl pkcs12 -in file.p12 -out apns_key.pem -nocerts -nodes
+        ```
+
+        The *topic* is the Bundle ID of the bridge recipient iOS application.
+        Since the cert needs to be tied directly to an application, the topic
+        is usually similar to "com.example.MyApplication".
+
+        :param cert_file: Path to the PEM formatted APNs certification file.
+        :type cert_file: str
+        :param key_file: Path to the PEM formatted APNs key file.
+        :type key_file: str
+        :param topic: The *Bundle ID* that identifies the assoc. iOS app.
+        :type topic: str
+        :param alt: Use the alternate APNs publication port (if 443 is blocked)
+        :type alt: bool
+        :param use_sandbox: Use the development sandbox
+        :type alt: bool
+        :param max_connections: Max number of pooled connections to use
+        :type max_connections: int
+        :param min_connections: Min number of pooled connections to use
+        :type min_connections: int
+        :param logger: Status logger
+        :type logger: logger
+        :param metrics: Metric recorder
+        :type metrics: autopush.metrics.IMetric
+        """
+        self.server = SERVER if not use_sandbox else SANDBOX
+        self.port = 443 if not alt else 2197
+        self.ssl_context = hyper.tls.init_context(cert=(cert_file, key_file))
+        self.log = logger
+        self.metrics = metrics
+        self.topic = topic
+        self._max_connections = max_connections
+        self._min_connections = min_connections
+        self.connections = deque(maxlen=max_connections)
+        if self.log:
+            self.log.info("Starting APNS connection")
+        # HTTP20Connection is not thread safe, so it's best to keep processing
+        # of the result in the same thread.
+        for i in xrange(0, min_connections):
```

Comment:
this pool logic here isn't quite right. this would create them and immediately throw them away (they're not added to self.connections)

max_connections doesn't seem to be really utilized either -- (the appendleft is guarded by len check vs min_connections)..